src/dmd/libomf.d

```diff
@@ -167,33 +167,36 @@ final class LibOMF : Library
 
     /*****************************************************************************/
 
-    void addSymbol(OmfObjModule* om, const(char)* name, int pickAny = 0)
+    void addSymbol(OmfObjModule* om, const(char)[] name, int pickAny = 0)
     {
         static if (LOG)
         {
-            printf("LibOMF::addSymbol(%s, %s, %d)\n", om.name.ptr, name, pickAny);
+            printf("LibOMF::addSymbol(%.*s, %.*s, %d)\n",
+                cast(int)om.name.length, om.name.ptr,
+                cast(int)name.length, name.ptr, pickAny);
         }
-        const namelen = strlen(name);
-        StringValue* s = tab.insert(name, namelen, null);
-        if (!s)
+        if (auto s = tab.insert(name, null))
+        {
+            auto os = new OmfObjSymbol();
+            os.name = name.toCStringThen!(cstr => strdup(cstr.ptr));
```

Comment:
This fails for two reasons:

1. toCStringThen returns a temporary on the stack containing the string. Then a pointer to that temporary gets stored in objsymbols, with disastrous results.
2. `name` is already 0-terminated. Allocating memory to 0-terminate it again is not an acceptable practice, and the way it is done adds a memory leak on top of that.

Reply:
I wasn't sure if that termination was guaranteed here. Reverted that part to a strdup(name.ptr) and added the assert recommended.

Reply:
@WalterBright : How are we sure `name` is `\0` terminated ?

Reply:
the assert() at the beginning of the function